hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/linux/resources/CGroupsV2HandlerImpl.java

```diff
@@ -0,0 +1,165 @@
+    if (type.equals(CGROUP2_FSTYPE)) {
+      return parseControllersFile(path);
+    }
+
+    return null;
+  }
+
+  @Override
+  protected void mountCGroupController(CGroupController controller) {
+    throw new UnsupportedOperationException("Mounting cgroup controllers is not supported in cgroup v2");
+  }
+
+  /**
+   * Parse the cgroup v2 controllers file to check the enabled controllers.
+   * @param cgroupPath path to the cgroup directory
+   * @return set of enabled and YARN supported controllers.
+   * @throws IOException if the file is not found or cannot be read
+   */
+  public Set<String> parseControllersFile(String cgroupPath) throws IOException {
+    File cgroupControllersFile = new File(cgroupPath + Path.SEPARATOR + CGROUP_CONTROLLERS_FILE);
+    if (!cgroupControllersFile.exists()) {
+      throw new IOException("No cgroup controllers file found in the directory specified: " +
+              cgroupPath);
+    }
+
+    String enabledControllers = FileUtils.readFileToString(cgroupControllersFile, StandardCharsets.UTF_8);
+    Set<String> validCGroups =
+            CGroupsHandler.CGroupController.getValidCGroups();
+    Set<String> controllerSet =
+            new HashSet<>(Arrays.asList(enabledControllers.split(" ")));
+    // Collect the valid subsystem names
+    controllerSet.retainAll(validCGroups);
+    if (controllerSet.isEmpty()) {
+      LOG.warn("The following cgroup directory doesn't contain any supported controllers: " +
+              cgroupPath);
+    }
+
+    return controllerSet;
+  }
+
+  /**
+   * Update the subtree_control file to enable subsequent container based cgroups to use the same controllers.
+   * @param yarnHierarchy path to the yarn cgroup under which the container cgroups will be created
+   * @throws ResourceHandlerException if the controllers file cannot be updated
+   */
+  @Override
+  protected void updateEnabledControllersInHierarchy(File yarnHierarchy) throws ResourceHandlerException {
+    PrintWriter pw = null;
+    try {
+      Set<String> enabledControllers = parseControllersFile(yarnHierarchy.getAbsolutePath());
+      if (enabledControllers.isEmpty()) {
+        throw new ResourceHandlerException("No valid controllers found in the cgroup hierarchy: " +
+                yarnHierarchy.getAbsolutePath());
+      }
+
+      File subtreeControlFile = new File(yarnHierarchy.getAbsolutePath()
+          + Path.SEPARATOR + CGROUP_SUBTREE_CONTROL_FILE);
+
+      Writer w = new OutputStreamWriter(Files.newOutputStream(subtreeControlFile.toPath()), StandardCharsets.UTF_8);
+      pw = new PrintWriter(w);
+      pw.write(String.join(" ", enabledControllers));
+
+    } catch (IOException e) {
+      throw new ResourceHandlerException("Failed to update the controllers file in the cgroup hierarchy: " +
+              yarnHierarchy.getAbsolutePath(), e);
+    } finally {
+      if (pw != null) {
+        pw.close();
+      }
+    }
```

Comment:
```suggestion
    try {
      Set<String> enabledControllers = parseControllersFile(yarnHierarchy.getAbsolutePath());
      if (enabledControllers.isEmpty()) {
        throw new ResourceHandlerException("No valid controllers found in the cgroup hierarchy: " +
                yarnHierarchy.getAbsolutePath());
      }

      File subtreeControlFile = new File(yarnHierarchy.getAbsolutePath()
          + Path.SEPARATOR + CGROUP_SUBTREE_CONTROL_FILE);

      Writer w = new OutputStreamWriter(Files.newOutputStream(subtreeControlFile.toPath()), StandardCharsets.UTF_8);
      try (PrintWriter pw = new PrintWriter(w)) {
        pw.write(String.join(" ", enabledControllers));
      }
    } catch (IOException e) {
      throw new ResourceHandlerException("Failed to update the controllers file in the cgroup hierarchy: " +
              yarnHierarchy.getAbsolutePath(), e);
    }
```

try with resources simplifies this code

Reply:
Thanks, applied.